### ISSUE TITLE
azuread_application dynamic optional_claims can cause a panic

### ISSUE THREAD

**Issue by dekimsey** [hashicorp/terraform-provider-azuread#522]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritise this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritise the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureAD Provider) Version

```
Terraform v1.0.0
on darwin_amd64
+ provider registry.terraform.io/hashicorp/azuread v1.6.0
```

### Affected Resource(s)

* `azuread_application`

### Terraform Configuration Files

```hcl

variable "optional_claims" {
  default = {}
}

resource "azuread_application" "this" {
  ...
  # https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims#optionalclaim-type
  optional_claims {
    dynamic "id_token" {
      for_each = lookup(var.optional_claims, "id_token", {})
      content {
        name = try(id_token.value.name, id_token.value)
        essential = try(id_token.value.essential, false)
      }
    }
    dynamic "access_token" {
      for_each = lookup(var.optional_claims, "access_token", {})
      content {
        name = try(id_token.value.name, id_token.value)
        essential = try(id_token.value.essential, false)
      }
    }
  }
}
```

### Debug Output

none available

### Panic Output

```

Stack trace from the terraform-provider-azuread_v1.6.0_x5 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 104 [running]:
github.com/hashicorp/terraform-provider-azuread/internal/services/applications.expandApplicationOptionalClaimsAad(0xc00088bc00, 0x1d69042)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/internal/services/applications/application_resource_aadgraph.go:533 +0x272
github.com/hashicorp/terraform-provider-azuread/internal/services/applications.applicationResourceUpdateAadGraph(0x1f645d8, 0xc00025ff80, 0xc00088bc00, 0x1c090e0, 0xc0001443c0, 0x1f645d8, 0xc00025ff80, 0xc000721a90)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/internal/services/applications/application_resource_aadgraph.go:268 +0x175a
github.com/hashicorp/terraform-provider-azuread/internal/services/applications.applicationResourceUpdate(0x1f645d8, 0xc00025ff80, 0xc00088bc00, 0x1c090e0, 0xc0001443c0, 0xc000721a90, 0x16b12aa, 0xc00028f880)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/internal/services/applications/application_resource.go:515 +0xde
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc0001d0b60, 0x1f64568, 0xc0008680c0, 0xc00088bc00, 0x1c090e0, 0xc0001443c0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:364 +0x17f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0001d0b60, 0x1f64568, 0xc0008680c0, 0xc0006bc1c0, 0xc00028f880, 0x1c090e0, 0xc0001443c0, 0x0, 0x0, 0x0, ...)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:464 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00000c210, 0x1f64568, 0xc0008680c0, 0xc00009c550, 0xc0008680c0, 0x1d10280, 0xc000746100)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc0005d4f60, 0x1f64610, 0xc0008680c0, 0xc0006bc0e0, 0xc0005d4f60, 0xc000746180, 0xc000763ba0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x1d10280, 0xc0005d4f60, 0x1f64610, 0xc000746180, 0xc00025e060, 0x0, 0x1f64610, 0xc000746180, 0xc000714a00, 0x2173)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001c2540, 0x1f6eff8, 0xc000219080, 0xc000880000, 0xc0006926c0, 0x25696a0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/google.golang.org/grpc/server.go:1217 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0001c2540, 0x1f6eff8, 0xc000219080, 0xc000880000, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/google.golang.org/grpc/server.go:1540 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc00003a2a0, 0xc0001c2540, 0x1f6eff8, 0xc000219080, 0xc000880000)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/google.golang.org/grpc/server.go:878 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-azuread/vendor/google.golang.org/grpc/server.go:876 +0x1fd

Error: The terraform-provider-azuread_v1.6.0_x5 plugin crashed!
```

### Expected Behavior

No optional claims are set.

### Actual Behavior

Crashes

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

This is likely due to the fact that the dynamic block is empty.

```
  # module.azure-app.azuread_application.this will be updated in-place
  ~ resource "azuread_application" "this" {
      + optional_claims {
        }
    }
```

**Workaround:** Wrapping the optional_claims in a dynamic call itself works around the behavior.

```
  dynamic "optional_claims" {
    for_each = length(var.optional_claims) > 0 ? ["placeholder"] : []
    content {
      dynamic "id_token" {
        ....
      }
      dynamic "access_token" {
        ...
      }
    }
  }
```

### References

None.

**Comment by manicminer** [comment]
Hi @dekimsey, thanks for reporting this crash scenario!

I've been able to reproduce this using azuread v1.6.0 without the msgraph beta. However, it looks like this does not occur when the msgraph beta is enabled, and similarly does not happen with a v2.0 candidate build.

Given that this only occurs when an empty `optional_claims` block is provided (which unfortunately seems to be the case inside a `dynamic` block), I don't think we'll be able to issue a backported fix for v1.6. Upgrading to v2.0 will resolve this, however, as a workaround in the interim you can enable the msgraph beta (which can be done for individual resources via provider aliases) and this will allow you to use an empty dynamic block.

**Comment by dekimsey** [comment]
I was also able to produce a work-around with nested dynamic blocks, so it's not a large concern for me now. I look forward to switching to v2.0 :)
